shell/browser/ui/file_dialog_linux.cc

```diff
@@ -0,0 +1,260 @@
+#include "base/run_loop.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "shell/browser/javascript_environment.h"
+#include "shell/browser/native_window_views.h"
+#include "shell/browser/ui/file_dialog.h"
+#include "shell/browser/ui/select_file_policy.h"
+#include "shell/common/gin_converters/callback_converter.h"
+#include "shell/common/gin_converters/file_path_converter.h"
+#include "ui/shell_dialogs/select_file_dialog.h"
+#include "ui/shell_dialogs/selected_file_info.h"
+
+namespace file_dialog {
+
+DialogSettings::DialogSettings() = default;
+DialogSettings::DialogSettings(const DialogSettings&) = default;
+DialogSettings::~DialogSettings() = default;
+
+namespace {
+
+ui::SelectFileDialog::Type GetDialogType(int properties) {
+  if (properties & OPEN_DIALOG_OPEN_FILE) {
+    return ui::SelectFileDialog::SELECT_OPEN_FILE;
+  } else if (properties & OPEN_DIALOG_OPEN_DIRECTORY) {
+    return ui::SelectFileDialog::SELECT_FOLDER;
+  } else if (properties & OPEN_DIALOG_MULTI_SELECTIONS) {
+    return ui::SelectFileDialog::SELECT_OPEN_MULTI_FILE;
+  }
+
+  return ui::SelectFileDialog::SELECT_OPEN_FILE;
+}
+
+// [save_label, show_overwrite_confirmation, show_hidden, ]
+std::tuple<std::string, bool, bool> GetSettingsTuple(
+    const DialogSettings& settings) {
+  bool show_overwrite_confirmation =
+      settings.properties & SAVE_DIALOG_SHOW_OVERWRITE_CONFIRMATION;
+  bool show_hidden = settings.properties & SAVE_DIALOG_SHOW_HIDDEN_FILES;
+
+  return std::make_tuple(settings.button_label, show_overwrite_confirmation,
+                         show_hidden);
+}
+
+ui::SelectFileDialog::FileTypeInfo GetFilterInfo(const Filters& filters) {
+  ui::SelectFileDialog::FileTypeInfo file_type_info;
+
+  for (const auto& filter : filters) {
+    auto [name, extension_group] = filter;
+    file_type_info.extension_description_overrides.push_back(
+        base::UTF8ToUTF16(name));
+
+    bool has_all_files_wildcard =
+        std::any_of(extension_group.begin(), extension_group.end(),
+                    [](std::string ext) { return ext == "*"; });
+    if (has_all_files_wildcard) {
+      file_type_info.include_all_files = true;
+    } else {
+      file_type_info.extensions.push_back(extension_group);
+    }
+  }
+
+  return file_type_info;
+}
+
+class FileChooserDialog : public ui::SelectFileDialog::Listener {
+ public:
+  enum class DialogType { OPEN, SAVE };
+
+  FileChooserDialog() {
+    dialog_ = ui::SelectFileDialog::Create(
+        this, std::make_unique<ElectronSelectFilePolicy>(nullptr));
+  }
+
+  ~FileChooserDialog() override {
+    // TODO(codebytere): anything we need to do here?
+  }
+
+  void RunSaveDialogImpl(const DialogSettings& settings) {
+    type_ = DialogType::SAVE;
+    ui::SelectFileDialog::FileTypeInfo file_info =
+        GetFilterInfo(settings.filters);
+    std::tuple<std::string, bool, bool> extra_settings =
+        GetSettingsTuple(settings);
```

Comment:
```suggestion
    auto extra_settings = GetExtraSettings(settings);
```

_(opinion)_  the function being called is visible only in this file, so I'd use `auto` here to futureproof this type.